File: protocol-generator/rust/src/cbor.rs
//! Hand-rolled canonical ECF CBOR codec.
//!
//! Owns the canonical layer in full (S1 `cbor_library = hand-rolled`), because
//! ECF's guarantees are stricter than any general CBOR library's "deterministic"
//! mode (profile [codec] A-RUST-001). Implements, byte-for-byte against the v1
//! conformance corpus:
//!
//! - Rule 1: shortest integer argument length (head minimization).
//! - Rule 2: map keys sorted by canonical encoded-key bytes (RFC 8949 §4.2.1;
//!   bytewise on encoded form = length-then-lex for this key space — matches
//!   the go oracle's `encodeMapCanonical`).
//! - Rule 3: definite lengths only; indefinite-length headers rejected on decode.
//! - Rule 4/4a: shortest-float ladder f64 -> f32 -> f16 with exact special bytes
//!   (NaN F97E00 / -0.0 F98000 / +Inf F97C00 / -Inf F9FC00), enforced on decode
//!   (a non-minimal float is non-canonical).
//! - Rule 5: duplicate map keys rejected on decode.
//! - §6.3: recursive major-type-6 (tag) rejection at ANY nesting depth.
//! - full uint64 / nint range (the int.10 .. [2^63, 2^64-1] band, via u64 carriers).
//! - raw-byte fidelity for byte strings.

use crate::error::{CodecError, Result};
use crate::value::{key_order, Key, Value};

// ───────────────────────── encode ─────────────────────────

/// Encode a value tree to canonical ECF bytes.
pub fn encode(v: &Value) -> Vec<u8> {
    let mut out = Vec::new();
    encode_into(v, &mut out);
    out
}

fn encode_into(v: &Value, out: &mut Vec<u8>) {
    match v {
        Value::Null => out.push(0xf6),
        Value::Bool(false) => out.push(0xf4),
        Value::Bool(true) => out.push(0xf5),
        Value::UInt(n) => encode_head(0, *n, out),
        Value::NInt(n) => encode_head(1, *n, out),
        Value::Float(f) => encode_float(*f, out),
        Value::Text(s) => {
            encode_head(3, s.len() as u64, out);
            out.extend_from_slice(s.as_bytes());
        }
        Value::Bytes(b) => {
            encode_head(2, b.len() as u64, out);
            out.extend_from_slice(b);
        }
        Value::Array(items) => {
            encode_head(4, items.len() as u64, out);
            for it in items {
                encode_into(it, out);
            }
        }
        Value::Map(entries) => {
            // Canonical Rule 2 sort by encoded key bytes (stable on the
            // already-encoded key form).
            let mut pairs: Vec<(Vec<u8>, &Value)> =
                entries.iter().map(|(k, val)| (k.encoded(), val)).collect();
            pairs.sort_by(|a, b| a.0.cmp(&b.0));
            encode_head(5, entries.len() as u64, out);
            for (kb, val) in pairs {
                out.extend_from_slice(&kb);
                encode_into(val, out);
            }
        }
    }
}

/// Emit a CBOR head: major type in the top 3 bits, shortest argument length
/// (RFC 8949 §4.2.1 Rule 1).
fn encode_head(major: u8, arg: u64, out: &mut Vec<u8>) {
    let m = major << 5;
    if arg < 24 {
        out.push(m | (arg as u8));
    } else if arg < 0x100 {
        out.push(m | 24);
        out.push(arg as u8);
    } else if arg < 0x1_0000 {
        out.push(m | 25);
        out.extend_from_slice(&(arg as u16).to_be_bytes());
    } else if arg < 0x1_0000_0000 {
        out.push(m | 26);
        out.extend_from_slice(&(arg as u32).to_be_bytes());
    } else {
        out.push(m | 27);
        out.extend_from_slice(&arg.to_be_bytes());
    }
}

/// Shortest-float encoding (Rule 4/4a). Try f16, then f32, then f64; the first
/// width that reproduces the exact f64 bit pattern (incl. -0.0 and NaN payload)
/// wins. Specials: +0.0/-0.0/Inf/NaN all collapse into f16.
fn encode_float(f: f64, out: &mut Vec<u8>) {
    if let Some(h) = f64_to_f16_exact(f) {
        out.push(0xf9);
        out.extend_from_slice(&h.to_be_bytes());
        return;
    }
    let s = f as f32;
    if (s as f64).to_bits() == f.to_bits() {
        out.push(0xfa);
        out.extend_from_slice(&s.to_bits().to_be_bytes());
        return;
    }
    out.push(0xfb);
    out.extend_from_slice(&f.to_bits().to_be_bytes());
}

/// Convert an f64 to a canonical f16 bit pattern IFF the half value reproduces
/// the f64 exactly. Returns `None` when f16 is lossy (so the ladder falls
/// through to f32/f64). Handles -0.0 (sign-preserving), the infinities, and the
/// canonical NaN payload 0x7e00.
fn f64_to_f16_exact(f: f64) -> Option<u16> {
    let bits = f.to_bits();
    let sign = ((bits >> 63) & 1) as u16;

    if f.is_nan() {
        // Canonical quiet NaN (Rule 4a). The corpus pins f97e00 for NaN.
        return Some(0x7e00);
    }
    if f.is_infinite() {
        return Some((sign << 15) | 0x7c00);
    }
    if f == 0.0 {
        // Covers +0.0 and -0.0; sign carries -0.0 -> 0x8000.
        return Some(sign << 15);
    }

    // Round-trip via the f32 first (half lives inside single's range/precision),
    // then half. The value survives only if half->f64 == original f64 exactly.
    let single = f as f32;
    if (single as f64).to_bits() != bits {
        return None; // not even f32-exact -> certainly not f16-exact
    }
    let half = f32_to_f16(single)?;
    if (f16_to_f64(half)).to_bits() == bits {
        Some(half)
    } else {
        None
    }
}

/// Encode an f32 into an f16 bit pattern when it is representable exactly as a
/// normal/subnormal half (no rounding). Returns `None` otherwise.
fn f32_to_f16(value: f32) -> Option<u16> {
    let x = value.to_bits();
    let sign = ((x >> 16) & 0x8000) as u16;
    let exp = ((x >> 23) & 0xff) as i32; // f32 biased exponent
    let mant = x & 0x007f_ffff; // 23-bit mantissa

    // Zero already handled by the caller; here value != 0.
    let unbiased = exp - 127;

    // Normal half range: unbiased exponent in [-14, 15].
    if (-14..=15).contains(&unbiased) {
        // f16 mantissa is 10 bits; the low 13 bits of the f32 mantissa must be
        // zero for an exact (non-rounding) conversion.
        if mant & 0x1fff != 0 {
            return None;
        }
        let half_exp = ((unbiased + 15) as u16) << 10;
        let half_mant = (mant >> 13) as u16;
        return Some(sign | half_exp | half_mant);
    }

    // Subnormal half range: 2^-24 .. 2^-15 (exact powers/sums representable).
    if (-24..-14).contains(&unbiased) {
        // Reconstruct the full significand (implicit leading 1 for normals).
        let full_mant = mant | 0x0080_0000; // 1.mant, 24 bits
        let shift = -14 - unbiased; // 1..10
                                    // Bits that would be lost on the shift into a 10-bit subnormal field.
        let total_shift = 13 + shift;
        let lost_mask = (1u32 << total_shift) - 1;
        if full_mant & lost_mask != 0 {
            return None;
        }
        let half_mant = (full_mant >> total_shift) as u16;
        return Some(sign | half_mant);
    }

    None
}

/// Decode an f16 bit pattern to f64 (exact; half is a subset of double).
fn f16_to_f64(h: u16) -> f64 {
    let sign = if (h >> 15) & 1 == 1 { -1.0 } else { 1.0 };
    let exp = ((h >> 10) & 0x1f) as i32;
    let mant = (h & 0x3ff) as f64;
    if exp == 0 {
        // Subnormal: 2^-14 * (mant / 1024).
        sign * 2f64.powi(-14) * (mant / 1024.0)
    } else if exp == 0x1f {
        if mant == 0.0 {
            sign * f64::INFINITY
        } else {
            f64::NAN
        }
    } else {
        // Normal: 2^(exp-15) * (1 + mant/1024).
        sign * 2f64.powi(exp - 15) * (1.0 + mant / 1024.0)
    }
}

// ───────────────────────── decode ─────────────────────────

/// Decode canonical ECF bytes to a value tree, rejecting any non-canonical
/// input (tags at any depth, indefinite lengths, non-minimal int/float,
/// duplicate or unsorted map keys, trailing data).
pub fn decode(bytes: &[u8]) -> Result<Value> {
    let mut d = Decoder { buf: bytes, pos: 0 };
    let v = d.value()?;
    if d.pos != d.buf.len() {
        return Err(CodecError::TrailingData);
    }
    Ok(v)
}

struct Decoder<'a> {
    buf: &'a [u8],
    pos: usize,
}

impl<'a> Decoder<'a> {
    fn take(&mut self, n: usize) -> Result<&'a [u8]> {
        if self.pos + n > self.buf.len() {
            return Err(CodecError::Truncated);
        }
        let s = &self.buf[self.pos..self.pos + n];
        self.pos += n;
        Ok(s)
    }

    fn byte(&mut self) -> Result<u8> {
        Ok(self.take(1)?[0])
    }

    /// Read a head argument for the given additional-info, enforcing minimal
    /// length (Rule 1). Returns the argument value.
    fn read_arg(&mut self, ai: u8) -> Result<u64> {
        match ai {
            0..=23 => Ok(ai as u64),
            24 => {
                let b = self.byte()?;
                if b < 24 {
                    return Err(CodecError::NonMinimalInt);
                }
                Ok(b as u64)
            }
            25 => {
                let b = self.take(2)?;
                let v = u16::from_be_bytes([b[0], b[1]]) as u64;
                if v < 0x100 {
                    return Err(CodecError::NonMinimalInt);
                }
                Ok(v)
            }
            26 => {
                let b = self.take(4)?;
                let v = u32::from_be_bytes([b[0], b[1], b[2], b[3]]) as u64;
                if v < 0x1_0000 {
                    return Err(CodecError::NonMinimalInt);
                }
                Ok(v)
            }
            27 => {
                let b = self.take(8)?;
                let v = u64::from_be_bytes([b[0], b[1], b[2], b[3], b[4], b[5], b[6], b[7]]);
                if v < 0x1_0000_0000 {
                    return Err(CodecError::NonMinimalInt);
                }
                Ok(v)
            }
            28..=30 => Err(CodecError::Malformed),
            31 => Err(CodecError::IndefiniteLength),
            _ => Err(CodecError::Malformed),
        }
    }

    fn value(&mut self) -> Result<Value> {
        let head = self.byte()?;
        let major = head >> 5;
        let ai = head & 0x1f;
        match major {
            0 => Ok(Value::UInt(self.read_arg(ai)?)),
            1 => Ok(Value::NInt(self.read_arg(ai)?)),
            2 => {
                if ai == 31 {
                    return Err(CodecError::IndefiniteLength);
                }
                let n = self.read_arg(ai)? as usize;
                Ok(Value::Bytes(self.take(n)?.to_vec()))
            }
            3 => {
                if ai == 31 {
                    return Err(CodecError::IndefiniteLength);
                }
                let n = self.read_arg(ai)? as usize;
                let s = self.take(n)?;
                let txt = std::str::from_utf8(s)
                    .map_err(|_| CodecError::Malformed)?
                    .to_string();
                Ok(Value::Text(txt))
            }
            4 => {
                if ai == 31 {
                    return Err(CodecError::IndefiniteLength);
                }
                let n = self.read_arg(ai)? as usize;
                let mut items = Vec::with_capacity(n);
                for _ in 0..n {
                    items.push(self.value()?);
                }
                Ok(Value::Array(items))
            }
            5 => {
                if ai == 31 {
                    return Err(CodecError::IndefiniteLength);
                }
                let n = self.read_arg(ai)? as usize;
                self.decode_map(n)
            }
            6 => {
                // §6.3: tags are forbidden at any nesting depth. Reject the
                // whole datum the moment a tag head is seen.
                Err(CodecError::TagRejected)
            }
            7 => self.decode_simple_or_float(ai),
            _ => Err(CodecError::Malformed),
        }
    }

    fn decode_map(&mut self, n: usize) -> Result<Value> {
        let mut entries: Vec<(Key, Value)> = Vec::with_capacity(n);
        let mut prev_key_enc: Option<Vec<u8>> = None;
        for _ in 0..n {
            let key_start = self.pos;
            let key = self.key()?;
            let key_enc = self.buf[key_start..self.pos].to_vec();
            // Rule 2 / Rule 5: keys must be strictly ascending by canonical
            // encoded bytes (also catches duplicates).
            if let Some(prev) = &prev_key_enc {
                match key_enc.as_slice().cmp(prev.as_slice()) {
                    std::cmp::Ordering::Less => return Err(CodecError::UnsortedKeys),
                    std::cmp::Ordering::Equal => return Err(CodecError::DuplicateKey),
                    std::cmp::Ordering::Greater => {}
                }
            }
            prev_key_enc = Some(key_enc);
            let val = self.value()?;
            entries.push((key, val));
        }
        Ok(Value::Map(entries))
    }

    /// Decode a value in key position into a [`Key`]. Tags/floats/containers are
    /// not valid keys in the corpus surface; reject anything unexpected.
    fn key(&mut self) -> Result<Key> {
        let head = self
            .buf
            .get(self.pos)
            .copied()
            .ok_or(CodecError::Truncated)?;
        let major = head >> 5;
        match major {
            0 => match self.value()? {
                Value::UInt(n) => Ok(Key::UInt(n)),
                _ => unreachable!(),
            },
            1 => match self.value()? {
                Value::NInt(n) => Ok(Key::NInt(n)),
                _ => unreachable!(),
            },
            2 => match self.value()? {
                Value::Bytes(b) => Ok(Key::Bytes(b)),
                _ => unreachable!(),
            },
            3 => match self.value()? {
                Value::Text(s) => Ok(Key::Text(s)),
                _ => unreachable!(),
            },
            6 => Err(CodecError::TagRejected),
            7 => {
                // Allow bool keys; reject float/null/simple keys.
                let ai = head & 0x1f;
                if ai == 20 || ai == 21 {
                    match self.value()? {
                        Value::Bool(b) => Ok(Key::Bool(b)),
                        _ => unreachable!(),
                    }
                } else {
                    Err(CodecError::Malformed)
                }
            }
            _ => Err(CodecError::Malformed),
        }
    }

    /// Major type 7: simple values + floats. ECF uses 0xf4/0xf5/0xf6; floats
    /// must be in shortest form (Rule 4 decode check).
    fn decode_simple_or_float(&mut self, ai: u8) -> Result<Value> {
        match ai {
            20 => Ok(Value::Bool(false)),
            21 => Ok(Value::Bool(true)),
            22 => Ok(Value::Null),
            23 => Err(CodecError::Malformed), // undefined (0xf7) not used in ECF
            25 => {
                // f16 — always shortest by construction; decode and accept.
                let b = self.take(2)?;
                let h = u16::from_be_bytes([b[0], b[1]]);
                Ok(Value::Float(f16_to_f64(h)))
            }
            26 => {
                let b = self.take(4)?;
                let bits = u32::from_be_bytes([b[0], b[1], b[2], b[3]]);
                let f = f32::from_bits(bits);
                // Reject if it would have fit in f16 (non-minimal float).
                if f64_to_f16_exact(f as f64).is_some() {
                    return Err(CodecError::NonMinimalFloat);
                }
                Ok(Value::Float(f as f64))
            }
            27 => {
                let b = self.take(8)?;
                let bits = u64::from_be_bytes([b[0], b[1], b[2], b[3], b[4], b[5], b[6], b[7]]);
                let f = f64::from_bits(bits);
                // Reject if a narrower form would reproduce it (non-minimal).
                if f64_to_f16_exact(f).is_some() {
                    return Err(CodecError::NonMinimalFloat);
                }
                let s = f as f32;
                if (s as f64).to_bits() == f.to_bits() {
                    return Err(CodecError::NonMinimalFloat);
                }
                Ok(Value::Float(f))
            }
            // ai 24 (1-byte simple), 28..=31 (reserved/indefinite), and simple
            // values 0..19 are all unused in ECF -> malformed.
            _ => Err(CodecError::Malformed),
        }
    }
}

/// Sort the entries of a map value into canonical order in place — a helper for
/// callers that build maps programmatically (the encoder sorts at emit time, so
/// this is only needed if one wants a pre-sorted in-memory form).
pub fn sort_map(entries: &mut [(Key, Value)]) {
    entries.sort_by(|a, b| key_order(&a.0, &b.0));
}

File: protocol-generator/julia/src/cbor.jl
# Entity Canonical Form (ECF) — hand-rolled canonical CBOR (profile [codec], A-JULIA-002).
#
# Why hand-rolled and not CBOR.jl: ECF (ENTITY-CBOR-ENCODING.md v1.5, RFC 8949 §4.2 with
# Entity clarifications) needs (a) length-then-lex map-key ordering, (b) shortest-float
# minimisation incl. f16, (c) recursive major-type-6 tag rejection on decode (N2),
# (d) full uint64/nint range. No Julia CBOR lib offers these, and a faithful ECF codec must
# OWN the canonical layer regardless (the A-005 pattern) — so a library buys nothing.
#
# THE HEADLINE MECHANISM (profile [idiom].multiple_dispatch_codec): `encode!` is a generic
# function dispatched by MULTIPLE DISPATCH on the Julia value TYPE — the type system makes
# CBOR major-type selection natural (String→mt3, Vector{UInt8}→mt2, Integer→mt0/1,
# AbstractFloat→mt7, Bool→mt7, Nothing→mt7). A-JULIA-006: branch exclusivity is verified —
# `String` is NOT an `AbstractVector`; a `Vector{UInt8}` is NOT a `String`; `Bool <: Integer`
# so the `Bool` method (more specific) wins over `Integer`; arrays are built as `Vector{Any}`
# so they never collide with byte-strings (`Vector{UInt8}`). No value shimmers across two
# branches.
module Cbor

export CborMap, encode, decode, decode_salvage, NonCanonicalECF, TruncatedInput, TagRejected

# ── error model (profile [error_model] = exceptions; custom <: Exception leaves) ──────────
abstract type EntityCoreError <: Exception end
struct NonCanonicalECF <: EntityCoreError; msg::String; end
struct TruncatedInput  <: EntityCoreError; msg::String; end
struct TagRejected     <: EntityCoreError; end
struct DuplicateKey    <: EntityCoreError; end
struct UnsupportedSimple <: EntityCoreError; ai::Int; end

# ── value model ───────────────────────────────────────────────────────────────────────────
# A CBOR map is an ORDERED pair list (NOT a Julia Dict): duplicate keys must be representable
# (to reject them, Rule 5) and byte-string keys must keep their type distinct from text keys.
# encode! sorts by encoded-key bytes, so insertion order is irrelevant to output.
struct CborMap
    pairs::Vector{Pair{Any,Any}}
end
CborMap() = CborMap(Pair{Any,Any}[])
CborMap(ps::AbstractVector) = CborMap(Pair{Any,Any}[Pair{Any,Any}(k, v) for (k, v) in ps])

# ── float16 exact decode (mirror of the encoder's round-trip check) ─────────────────────────
function half_to_double(h::UInt16)::Float64
    # Julia's Float16 is native: reinterpret the bits and widen. Handles ±0, subnormal,
    # inf and NaN exactly per IEEE-754 binary16→binary64.
    return Float64(reinterpret(Float16, h))
end

# ── big-endian argument emit ────────────────────────────────────────────────────────────────
@inline function be!(out::Vector{UInt8}, v::UInt64, nbytes::Int)
    i = nbytes
    while i > 0
        i -= 1
        push!(out, UInt8((v >> (8 * i)) & 0xff))
    end
    return nothing
end

# Emit a CBOR head: major type (0..7) << 5 | minimal-length unsigned argument (RFC 8949 §4.2.1).
@inline function addhead!(out::Vector{UInt8}, major::UInt8, arg::UInt64)
    mt = major << 5
    if arg < 0x18                      # < 24  → immediate
        push!(out, mt | UInt8(arg))
    elseif arg < 0x100                 # 1-byte argument
        push!(out, mt | 0x18); be!(out, arg, 1)
    elseif arg < 0x10000               # 2-byte argument
        push!(out, mt | 0x19); be!(out, arg, 2)
    elseif arg < 0x100000000           # 4-byte argument
        push!(out, mt | 0x1a); be!(out, arg, 4)
    else                               # 8-byte argument
        push!(out, mt | 0x1b); be!(out, arg, 8)
    end
    return nothing
end

# ── canonical map-key ordering: encoded-length ascending, then bytewise-lex ascending ───────
@inline function canon_less(a::Vector{UInt8}, b::Vector{UInt8})::Bool
    la, lb = length(a), length(b)
    la != lb && return la < lb
    return a < b            # Julia's Vector{UInt8} isless is bytewise-lexicographic
end

# ── shortest-float ladder (Rule 4): f16 ← f32 ← f64, native Float16/Float32 ──────────────────
# NaN canonicalises to f9 7e00 (Rule 4a). Otherwise EMIT the narrowest form that round-trips
# BIT-EXACTLY back to the f64 (bit compare via reinterpret so -0.0 ≠ 0.0). Float16 is a native
# Julia type, so the f16 leg needs no hand-rolled half-float packer — but the ladder DECISION
# (does it round-trip?) is hand-made regardless.
function encode_float!(out::Vector{UInt8}, x::Float64)
    if isnan(x)
        push!(out, 0xf9); be!(out, 0x7e00 % UInt64, 2); return nothing
    end
    xb = reinterpret(UInt64, x)
    h = Float16(x)
    if reinterpret(UInt64, Float64(h)) == xb
        push!(out, 0xf9); be!(out, UInt64(reinterpret(UInt16, h)), 2); return nothing
    end
    s = Float32(x)
    if reinterpret(UInt64, Float64(s)) == xb
        push!(out, 0xfa); be!(out, UInt64(reinterpret(UInt32, s)), 4); return nothing
    end
    push!(out, 0xfb); be!(out, xb, 8); return nothing
end

# ── encode! — multiple dispatch on the value type (the headline mechanism) ──────────────────
# Bool BEFORE Integer: `Bool <: Integer`, and the Bool method is more specific, so dispatch
# selects it (A-JULIA-006). Nothing → mt7 null. Wire integer carrier is UInt64 (A-JULIA-008):
# the mt0/mt1 argument is exact over [0, 2^64-1]; negative values fold via -1-n in BigInt space
# then narrow, so no fixed-width overflow at the 2^63/2^64 boundary.
encode!(out::Vector{UInt8}, ::Nothing) = (push!(out, 0xf6); nothing)
encode!(out::Vector{UInt8}, b::Bool)   = (push!(out, b ? 0xf5 : 0xf4); nothing)

function encode!(out::Vector{UInt8}, n::Integer)
    if n < 0
        addhead!(out, 0x01, UInt64(-1 - big(n)))      # mt1: argument = -1 - n
    else
        addhead!(out, 0x00, UInt64(n))                # mt0: argument = n  (exact to 2^64-1)
    end
    return nothing
end

encode!(out::Vector{UInt8}, x::AbstractFloat) = encode_float!(out, Float64(x))

function encode!(out::Vector{UInt8}, b::AbstractVector{UInt8})   # mt2 byte string
    addhead!(out, 0x02, UInt64(length(b)))
    append!(out, b)
    return nothing
end

function encode!(out::Vector{UInt8}, s::AbstractString)          # mt3 text string
    cu = codeunits(s)                                # UTF-8 bytes; length is BYTE length
    addhead!(out, 0x03, UInt64(length(cu)))          # A-JULIA-007: ncodeunits, NEVER length(s)
    append!(out, cu)
    return nothing
end

function encode!(out::Vector{UInt8}, v::AbstractVector)          # mt4 array (Vector{Any} etc.)
    addhead!(out, 0x04, UInt64(length(v)))
    for it in v
        encode!(out, it)
    end
    return nothing
end

function encode!(out::Vector{UInt8}, m::CborMap)                 # mt5 map (length-then-lex)
    keyed = Tuple{Vector{UInt8},Any}[]
    for p in m.pairs
        kb = UInt8[]
        encode!(kb, p.first)
        push!(keyed, (kb, p.second))
    end
    sort!(keyed; lt = (a, b) -> canon_less(a[1], b[1]))
    for i in 2:length(keyed)
        keyed[i-1][1] == keyed[i][1] && throw(DuplicateKey())    # Rule 5
    end
    addhead!(out, 0x05, UInt64(length(m.pairs)))
    for (kb, val) in keyed
        append!(out, kb)
        encode!(out, val)
    end
    return nothing
end

"""Encode `v` to a fresh canonical-ECF byte vector."""
function encode(v)::Vector{UInt8}
    out = UInt8[]
    encode!(out, v)
    return out
end

# ── decode (rejects tags at any depth, indefinite lengths, trailing bytes) ──────────────────
mutable struct Decoder
    s::Vector{UInt8}
    pos::Int          # 1-based
    # keep_tags makes decode_item yield the tag's INNER item instead of throwing
    # TagRejected. It exists for ONE caller — decode_salvage — and is never set on
    # the strict path. See decode_salvage for why this is not a weakening of §6.3.
    keep_tags::Bool
end
Decoder(s::Vector{UInt8}, pos::Int) = Decoder(s, pos, false)

@inline function need(d::Decoder, k::Int)
    d.pos + k - 1 > length(d.s) && throw(TruncatedInput("need $k byte(s) at pos $(d.pos)"))
    return nothing
end
@inline function readbyte(d::Decoder)::UInt8
    need(d, 1); c = d.s[d.pos]; d.pos += 1; return c
end
@inline function take(d::Decoder, k::Int)::Vector{UInt8}
    need(d, k); r = d.s[d.pos:d.pos+k-1]; d.pos += k; return r
end
@inline function be(d::Decoder, k::Int)::UInt64
    need(d, k); v = UInt64(0)
    for _ in 1:k
        v = (v << 8) | UInt64(d.s[d.pos]); d.pos += 1
    end
    return v
end
@inline function read_arg(d::Decoder, ai::UInt8)::UInt64
    ai < 0x18 && return UInt64(ai)
    ai == 0x18 && return be(d, 1)
    ai == 0x19 && return be(d, 2)
    ai == 0x1a && return be(d, 4)
    ai == 0x1b && return be(d, 8)
    throw(NonCanonicalECF("indefinite/reserved length argument (ai=$ai)"))   # 28..31
end

function decode_item(d::Decoder)
    ib = readbyte(d)
    major = ib >> 5
    ai = ib & 0x1f
    if major == 0x00
        arg = read_arg(d, ai)
        return arg <= UInt64(typemax(Int64)) ? Int64(arg) : arg      # widen only above i64-max
    elseif major == 0x01
        n = read_arg(d, ai)                                          # value = -1 - n
        return n < UInt64(typemax(Int64)) ? Int64(-1 - Int64(n)) : (-1 - Int128(n))
    elseif major == 0x02
        return take(d, Int(read_arg(d, ai)))                         # Vector{UInt8}
    elseif major == 0x03
        return String(take(d, Int(read_arg(d, ai))))                # UTF-8 text
    elseif major == 0x04
        len = Int(read_arg(d, ai))
        items = Vector{Any}(undef, len)
        for i in 1:len
            items[i] = decode_item(d)
        end
        return items
    elseif major == 0x05
        len = Int(read_arg(d, ai))
        ps = Vector{Pair{Any,Any}}(undef, len)
        for i in 1:len
            k = decode_item(d)
            v = decode_item(d)
            ps[i] = Pair{Any,Any}(k, v)
        end
        return CborMap(ps)
    elseif major == 0x06
        d.keep_tags || throw(TagRejected())                         # N2: any tag, any depth
        # Salvage path only (decode_salvage): consume the tag head and yield the item
        # it wrapped, so the caller can locate the request_id and SIGNAL the rejection.
        # The frame is still rejected — the tag is never interpreted and the value
        # never reaches an entity.
        read_arg(d, ai)
        return decode_item(d)
    else # major == 0x07
        if ai == 0x14
            return false
        elseif ai == 0x15
            return true
        elseif ai == 0x16
            return nothing
        elseif ai == 0x19
            return half_to_double(UInt16(be(d, 2)))
        elseif ai == 0x1a
            return Float64(reinterpret(Float32, UInt32(be(d, 4))))
        elseif ai == 0x1b
            return reinterpret(Float64, be(d, 8))
        else
            throw(UnsupportedSimple(Int(ai)))
        end
    end
end

"""Decode a single top-level ECF item; rejects trailing bytes."""
function decode(s::AbstractVector{UInt8})
    d = Decoder(Vector{UInt8}(s), 1, false)
    v = decode_item(d)
    d.pos != length(d.s) + 1 && throw(NonCanonicalECF("trailing bytes after top-level item"))
    return v
end

"""
Decode `s` for the sole purpose of REPORTING a rejection, not of accepting one.
Identical to `decode` except that a major-type-6 tag yields the item it wrapped
instead of throwing `TagRejected`.

Why this exists (§6.3, a conformance requirement rather than a convenience): the tag
rule is *"Implementations MUST reject any received protocol frame containing a CBOR
tag on a data field. Rejection returns `400 non_canonical_ecf`."* Rejecting by
dropping the frame on the floor satisfies the first sentence and violates the second
— the peer owes the sender a status, and §4.9(c) deliver-or-signal says the same from
the other direction. But the status must ride a response correlated by `request_id`,
and the strict decoder cannot reach the request_id in a frame it refuses to parse.
This recovers exactly that much and nothing more.

This is NOT a weakening of the tag reject. The frame stays rejected: the value this
returns is never converted to an Entity, never stored, never forwarded and never
interpreted, so §6.3's MUST NOT silently strip / MUST NOT preserve / MUST NOT attempt
to interpret all still hold. The strict `decode` path that every real ingestion route
uses is unchanged, which is what keeps the `tag_reject` wire-conformance vectors
meaningful.
"""
function decode_salvage(s::AbstractVector{UInt8})
    d = Decoder(Vector{UInt8}(s), 1, true)
    v = decode_item(d)
    d.pos != length(d.s) + 1 && throw(NonCanonicalECF("trailing bytes after top-level item"))
    return v
end

end # module Cbor
